lightning/src/util/persist.rs

```diff
@@ -250,3 +261,942 @@ where
+///
+/// Updates would be stored as follows (with `/` delimiting
+/// namespace/subnamespace/key):
+///
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000002`
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000003`
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000004`
+/// ... and so on.
+///
+/// # Reading channel state from storage
+///
+/// Channel state is reconstructed by calling
+/// [`read_channel_monitors_with_updates`]. Please see that function's
+/// documentation for more information.
+///
+/// # Pruning stale channel updates
+///
+/// TODO(domz)
+pub struct MonitorUpdatingPersister<K: Deref, L: Deref>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	kv_store: K,
+	logger: L,
+	maximum_pending_updates: u64,
+	last_monitor_update_id: RefCell<HashMap<MonitorName, u64>>,
+}
+
+#[allow(dead_code)]
+impl<K: Deref, L: Deref> MonitorUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	/// Read channel monitors, along with any stored updates for them, applies
+	/// the updates, and returns the fully updated monitors in the resulting
+	/// vector.
+	///
+	/// # EXTREMELY IMPORTANT
+	/// It is **EXTREMELY IMPORTANT** that your [`KVStore`] implementation use
+	/// the [`std::io::error::ErrorKind::NotFound`] variant correctly: that is,
+	/// when a file is not found, and _only_ in that circumstance (not when
+	/// there is really a permissions error, for example). This is because this
+	/// function does not list updates. Instead, it lists all monitors, and then
+	/// using their stored [`update_id`], synthesizes update storage keys and
+	/// tries them in sequence until one is not found. All _other_ errors will
+	/// be bubbled up in this function's [`io::Result`].
+	pub fn read_channel_monitors_with_updates<
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+		B: Deref,
+		F: Deref + Clone,
+	>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		broadcaster: &B,
+		fee_estimator: F,
+	) -> io::Result<
+		Vec<(
+			BlockHash,
+			ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+		)>,
+	>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let mut res = Vec::new();
+		// for each monitor...
+		for monitor_name in self.list_monitor_names()? {
+			// ...parse the monitor
+			let (bh, monitor) = self.deserialize_monitor(
+				entropy_source.clone(),
+				signer_provider.clone(),
+				&monitor_name,
+			)?;
+			// ...parse and apply the updates with an id higher than the monitor.
+			let mut update_id = monitor.get_latest_update_id();
+			self.last_monitor_update_id
+				.borrow_mut()
+				.insert(monitor_name.clone(), update_id);
+			loop {
```

Comment:
I think both ways express their respective designs equally well; I think they're actually different designs. If we put a finite range here, this goes from:

> read monitors until we don't find one

to:

> read monitors until we don't find one OR we read `maximum_pending_updates`

I think the second design would open us up to penalty transactions if there are ever updates above the end of the loop. I think we're _pretty_ confident our design below cannot do that. But, as we intend for this number to be user-configurable, they could lower it between builds! Now the inputs to the function have changed. In that scenario, unless they very carefully apply updates out of band, etc., they will stop short of stored fresh updates and get penalized, I think?

Reply:
I had the same reservation, actually. I started playing around with a boundary, like maybe having a u64 counter that breaks/errors/panics (not sure which) when saturated. Just didn't add it because I'm not sure it's necessary, and I wonder if it could mask a problem. An infinite loop, while annoying, might be preferable to that.